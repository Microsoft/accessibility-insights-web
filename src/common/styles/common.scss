// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import 'high-contrast-override.scss';
@import 'colors.scss';
@import 'fonts.scss';

$focus-ring-color: $neutral-100;

@mixin set-focus-ring-styles($property: 'border') {
    #{$property}: 1px solid $focus-ring-color;
    @include ms-high-contrast-override {
        #{$property}: 1px dashed windowtext !important;
    }
}

.ms-Fabric.is-focusVisible .ms-Nav-group .ms-nav-linkButton:focus::after {
    @include set-focus-ring-styles('border');
}

button::after {
    content: '';
    position: absolute;
}

.ms-Button.ms-Button--primary:not(.is-disabled) {
    background: $communication-primary;

    &:hover {
        background: $communication-shade-10;
    }
}

.ms-fontColor-neutralPrimary,
.ms-fontColor-neutralPrimary--hover:hover {
    color: $neutral-100 !important;
}

.ms-Fabric {
    color: $neutral-100;
}

.generic-toggle-component {
    margin-bottom: 4vh;
    .toggle-container {
        margin-bottom: 1vh;
        .toggle-name {
            display: inline-block;
            font-size: 17px;
            font-weight: bold;
        }

        .toggle {
            float: right;
        }
    }
}

.high-contrast-theme {
    .insights-link {
        &:hover {
            text-decoration: underline;
        }
    }
    .is-selected {
        .status-icon {
            color: $high-contrast-icon-color !important;
        }
    }
}

.insights-link {
    // marking important to override office fabric style
    color: $link !important;
    text-decoration: none;

    &:hover {
        color: $link-hover !important;
        text-decoration: underline;
    }
}

.ms-Spinner {
    .ms-Spinner-circle {
        border-top-color: $spinner-text;
    }

    .ms-Spinner-label {
        color: $spinner-text;
    }
}

$normalTitleFontWeight: 300 !default;

$pivotItemLeftBorderWidth: 3px !default;
$pivotItemBorderStyle: solid !default;

$detailsViewHeaderBarHeight: 40px !default;
$detailsViewNavPivotsHeight: 41px !default;
$detailsViewCommandBarHeight: 40px !default;
$detailsViewCommandBarBorderHeight: 1px !default;
$detailsViewLeftNavWidth: 200px !default;
$detailsViewTotalHeaderHeight: (#{$detailsViewCommandBarBorderHeight} + #{$detailsViewCommandBarHeight} + #{$detailsViewHeaderBarHeight});

$fastPassRightPanelMarginRight: 24px !default;
$fastPassRightPanelMarginLeft: 24px !default;
$fastPassRightPanelMarginTop: 24px !default;

@mixin ellipsedText {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}
