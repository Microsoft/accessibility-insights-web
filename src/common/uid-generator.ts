// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as uuid4 from 'uuid/v4';

export type UUIDGenerator = () => string;

export let generateUID: UUIDGenerator = uuid4;
