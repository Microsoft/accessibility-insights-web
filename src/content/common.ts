// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as react from 'react';

import { link } from './link';
import { ContentCreator } from '../views/content/content-page';
import { GuidanceTitle as gt } from '../views/content/guidance-title';

export const create = ContentCreator(link);
export const React = react;
export const GuidanceTitle = gt;
