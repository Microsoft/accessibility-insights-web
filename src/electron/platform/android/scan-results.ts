// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ScreenshotData } from 'common/types/store-data/unified-data-interface';

export interface RuleResultsData {
    axeViewId: string;
    ruleId: string;
    status: string;
    props: any;
}

export interface ViewElementData {
    axeViewId: string;
    boundsInScreen: BoundingRectangle;
    className: string;
    contentDescription: string;
    text: string;
    children: ViewElementData[];
}

export interface BoundingRectangle {
    bottom: number;
    left: number;
    right: number;
    top: number;
}

export class ScanResults {
    constructor(readonly rawData: any) {}

    public get deviceName(): string {
        try {
            return this.rawData.axeContext.axeDevice.name;
        } catch {
            return null;
        }
    }

    public get appIdentifier(): string {
        try {
            return this.rawData.axeContext.axeMetaData.appIdentifier;
        } catch {
            return null;
        }
    }

    public get viewElementTree(): ViewElementData {
        try {
            return this.rawData.axeContext.axeView;
        } catch {
            return null;
        }
    }

    public get ruleResults(): RuleResultsData[] {
        try {
            const results = this.rawData.axeRuleResults;
            return results || [];
        } catch {
            return [];
        }
    }

    public get axeVersion(): string {
        try {
            return this.rawData.axeContext.axeMetaData.axeVersion;
        } catch {
            return 'no-version';
        }
    }

    public get screenshot(): ScreenshotData {
        try {
            return {
                base64PngData: this.rawData.axeContext.screenshot,
            };
        } catch {
            return null;
        }
    }
}
