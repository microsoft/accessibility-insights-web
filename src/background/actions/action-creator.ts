// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { CardSelectionActions } from 'background/actions/card-selection-actions';
import { SidePanelActions } from 'background/actions/side-panel-actions';
import { UnifiedScanResultActions } from 'background/actions/unified-scan-result-actions';
import { TestMode } from 'common/configs/test-mode';
import { VisualizationConfigurationFactory } from 'common/configs/visualization-configuration-factory';
import * as TelemetryEvents from 'common/extension-telemetry-events';
import { Logger } from 'common/logging/logger';
import { getStoreStateMessage, Messages } from 'common/messages';
import { NotificationCreator } from 'common/notification-creator';
import { StoreNames } from 'common/stores/store-names';
import { VisualizationType } from 'common/types/visualization-type';
import { ScanCompletedPayload } from 'injected/analyzers/analyzer';
import { DictionaryNumberTo } from 'types/common-types';
import { VisualizationActions } from '../actions/visualization-actions';
import { VisualizationScanResultActions } from '../actions/visualization-scan-result-actions';
import { ExtensionDetailsViewController } from '../extension-details-view-controller';
import { Interpreter } from '../interpreter';
import { TargetTabController } from '../target-tab-controller';
import { TelemetryEventHandler } from '../telemetry/telemetry-event-handler';
import { ActionHub } from './action-hub';
import {
    AddTabbedElementPayload,
    BaseActionPayload,
    OnDetailsViewOpenPayload,
    OnDetailsViewPivotSelected,
    RescanVisualizationPayload,
    ToggleActionPayload,
    VisualizationTogglePayload,
} from './action-payloads';
import { InspectActions } from './inspect-actions';

const visualizationMessages = Messages.Visualizations;

export class ActionCreator {
    private visualizationActions: VisualizationActions;
    private visualizationScanResultActions: VisualizationScanResultActions;
    private adHocTestTypeToTelemetryEvent: DictionaryNumberTo<string> = {
        [VisualizationType.Color]: TelemetryEvents.COLOR_TOGGLE,
        [VisualizationType.Headings]: TelemetryEvents.HEADINGS_TOGGLE,
        [VisualizationType.Issues]: TelemetryEvents.AUTOMATED_CHECKS_TOGGLE,
        [VisualizationType.Landmarks]: TelemetryEvents.LANDMARKS_TOGGLE,
        [VisualizationType.TabStops]: TelemetryEvents.TABSTOPS_TOGGLE,
        [VisualizationType.NeedsReview]: TelemetryEvents.NEEDS_REVIEW_TOGGLE,
    };
    private inspectActions: InspectActions;
    private cardSelectionActions: CardSelectionActions;
    private unifiedScanResultActions: UnifiedScanResultActions;
    private sidePanelActions: SidePanelActions;

    constructor(
        private readonly interpreter: Interpreter,
        readonly actionHub: ActionHub,
        private readonly detailsViewController: ExtensionDetailsViewController,
        private readonly telemetryEventHandler: TelemetryEventHandler,
        private readonly notificationCreator: NotificationCreator,
        private readonly visualizationConfigurationFactory: VisualizationConfigurationFactory,
        private readonly targetTabController: TargetTabController,
        private readonly logger: Logger,
    ) {
        this.visualizationActions = actionHub.visualizationActions;
        this.visualizationScanResultActions = actionHub.visualizationScanResultActions;
        this.inspectActions = actionHub.inspectActions;
        this.cardSelectionActions = actionHub.cardSelectionActions;
        this.unifiedScanResultActions = actionHub.scanResultActions;
        this.sidePanelActions = actionHub.sidePanelActions;
    }

    public registerCallbacks(): void {
        this.interpreter.registerTypeToPayloadCallback(
            visualizationMessages.Common.Toggle,
            this.onVisualizationToggle,
        );
        this.interpreter.registerTypeToPayloadCallback(
            visualizationMessages.Common.ScanCompleted,
            this.onAdHocScanCompleted,
        );
        this.interpreter.registerTypeToPayloadCallback(
            visualizationMessages.Common.ScrollRequested,
            this.onScrollRequested,
        );
        this.interpreter.registerTypeToPayloadCallback(
            visualizationMessages.Common.RescanVisualization,
            this.onRescanVisualization,
        );

        this.interpreter.registerTypeToPayloadCallback(
            visualizationMessages.Issues.UpdateFocusedInstance,
            this.onUpdateFocusedInstance,
        );

        this.interpreter.registerTypeToPayloadCallback(
            getStoreStateMessage(StoreNames.VisualizationStore),
            this.getVisualizationToggleCurrentState,
        );
        this.interpreter.registerTypeToPayloadCallback(
            getStoreStateMessage(StoreNames.VisualizationScanResultStore),
            this.getScanResultsCurrentState,
        );

        this.interpreter.registerTypeToPayloadCallback(
            visualizationMessages.TabStops.TabbedElementAdded,
            this.onTabbedElementAdded,
        );
        this.interpreter.registerTypeToPayloadCallback(
            visualizationMessages.TabStops.RecordingCompleted,
            this.onRecordingCompleted,
        );
        this.interpreter.registerTypeToPayloadCallback(
            visualizationMessages.TabStops.TerminateScan,
            this.onRecordingTerminated,
        );

        this.interpreter.registerTypeToPayloadCallback(
            visualizationMessages.DetailsView.Open,
            this.onDetailsViewOpen,
        );
        this.interpreter.registerTypeToPayloadCallback(
            visualizationMessages.DetailsView.Select,
            this.onPivotChildSelected,
        );
        this.interpreter.registerTypeToPayloadCallback(
            visualizationMessages.DetailsView.PivotSelect,
            this.onDetailsViewPivotSelected,
        );
        this.interpreter.registerTypeToPayloadCallback(
            visualizationMessages.DetailsView.Close,
            this.onDetailsViewClosed,
        );
        this.interpreter.registerTypeToPayloadCallback(
            Messages.Assessment.AssessmentScanCompleted,
            this.onAssessmentScanCompleted,
        );
        this.interpreter.registerTypeToPayloadCallback(
            Messages.Assessment.StartOverTest,
            this.onStartOver,
        );
        this.interpreter.registerTypeToPayloadCallback(
            Messages.Assessment.CancelStartOver,
            this.onCancelStartOver,
        );
        this.interpreter.registerTypeToPayloadCallback(
            Messages.Assessment.StartOverAllAssessments,
            this.onStartOverAllAssessments,
        );
        this.interpreter.registerTypeToPayloadCallback(
            Messages.Assessment.CancelStartOverAllAssessments,
            this.onCancelStartOverAllAssessments,
        );
        this.interpreter.registerTypeToPayloadCallback(
            Messages.Assessment.EnableVisualHelper,
            this.onEnableVisualHelper,
        );
        this.interpreter.registerTypeToPayloadCallback(
            Messages.Assessment.DisableVisualHelperForTest,
            this.onDisableVisualHelpersForTest,
        );
        this.interpreter.registerTypeToPayloadCallback(
            Messages.Assessment.DisableVisualHelper,
            this.onDisableVisualHelper,
        );
        this.interpreter.registerTypeToPayloadCallback(
            Messages.Assessment.EnableVisualHelperWithoutScan,
            this.onEnableVisualHelperWithoutScan,
        );

        this.interpreter.registerTypeToPayloadCallback(
            Messages.Inspect.SetHoveredOverSelector,
            this.onSetHoveredOverSelector,
        );
    }

    private onEnableVisualHelperWithoutScan = (payload: ToggleActionPayload): void => {
        this.visualizationActions.enableVisualizationWithoutScan.invoke(payload);
    };

    private onEnableVisualHelper = (payload: ToggleActionPayload): void => {
        this.visualizationActions.enableVisualization.invoke(payload);
    };

    private onDisableVisualHelpersForTest = (payload: ToggleActionPayload): void => {
        this.visualizationActions.disableVisualization.invoke(payload.test);
    };

    private onDisableVisualHelper = (payload: ToggleActionPayload): void => {
        const eventName = TelemetryEvents.DISABLE_VISUAL_HELPER;
        this.telemetryEventHandler.publishTelemetry(eventName, payload);
        this.visualizationActions.disableVisualization.invoke(payload.test);
    };

    private onStartOver = (payload: ToggleActionPayload): void => {
        const eventName = TelemetryEvents.START_OVER_TEST;
        this.telemetryEventHandler.publishTelemetry(eventName, payload);
        this.visualizationActions.disableVisualization.invoke(payload.test);
    };

    private onCancelStartOver = (payload: BaseActionPayload): void => {
        const eventName = TelemetryEvents.CANCEL_START_OVER_TEST;
        this.telemetryEventHandler.publishTelemetry(eventName, payload);
    };

    private onStartOverAllAssessments = (payload: ToggleActionPayload): void => {
        const eventName = TelemetryEvents.START_OVER_ASSESSMENT;
        this.telemetryEventHandler.publishTelemetry(eventName, payload);
        this.visualizationActions.disableAssessmentVisualizations.invoke(null);
    };

    private onCancelStartOverAllAssessments = (payload: BaseActionPayload): void => {
        const eventName = TelemetryEvents.CANCEL_START_OVER_ASSESSMENT;
        this.telemetryEventHandler.publishTelemetry(eventName, payload);
    };

    private onDetailsViewClosed = (): void => {
        this.visualizationActions.disableAssessmentVisualizations.invoke(null);
    };

    private onAssessmentScanCompleted = async (
        payload: ScanCompletedPayload<any>,
        tabId: number,
    ): Promise<void> => {
        const eventName = TelemetryEvents.ASSESSMENT_SCAN_COMPLETED;
        this.telemetryEventHandler.publishTelemetry(eventName, payload);
        this.visualizationActions.scanCompleted.invoke(null);
        this.notificationCreator.createNotificationByVisualizationKey(
            payload.selectorMap,
            payload.key,
            payload.testType,
            payload.scanIncompleteWarnings,
        );
        await this.targetTabController.showTargetTab(tabId, payload.testType, payload.key);
    };

    private onTabbedElementAdded = (payload: AddTabbedElementPayload): void => {
        this.visualizationScanResultActions.addTabbedElement.invoke(payload);
    };

    private onRecordingCompleted = (payload: BaseActionPayload): void => {
        this.telemetryEventHandler.publishTelemetry(
            TelemetryEvents.TABSTOPS_RECORDING_COMPLETE,
            payload,
        );
    };

    private onRecordingTerminated = (payload: BaseActionPayload): void => {
        this.visualizationScanResultActions.disableTabStop.invoke(payload);
    };

    private onUpdateFocusedInstance = (payload: string[]): void => {
        this.visualizationActions.updateFocusedInstance.invoke(payload);
    };

    private onAdHocScanCompleted = async (
        payload: ScanCompletedPayload<any>,
        tabId: number,
    ): Promise<void> => {
        const telemetryEventName = TelemetryEvents.ADHOC_SCAN_COMPLETED;
        this.telemetryEventHandler.publishTelemetry(telemetryEventName, payload);
        this.visualizationScanResultActions.scanCompleted.invoke(payload);
        this.visualizationActions.scanCompleted.invoke(null);
        this.notificationCreator.createNotificationByVisualizationKey(
            payload.selectorMap,
            payload.key,
            payload.testType,
            payload.scanIncompleteWarnings,
        );
        await this.targetTabController.showTargetTab(tabId, payload.testType, payload.key);
    };

    private onScrollRequested = (): void => {
        this.visualizationActions.scrollRequested.invoke(null);
        this.cardSelectionActions.resetFocusedIdentifier.invoke(null);
    };

    private onDetailsViewOpen = async (
        payload: OnDetailsViewOpenPayload,
        tabId: number,
    ): Promise<void> => {
        if (this.shouldEnableToggleOnDetailsViewOpen(payload.detailsViewType)) {
            this.enableToggleOnDetailsViewOpen(payload.detailsViewType, tabId);
        }

        await this.onPivotChildSelected(payload, tabId);
    };

    private shouldEnableToggleOnDetailsViewOpen(visualizationType: VisualizationType): boolean {
        return (
            visualizationType != null &&
            visualizationType !== VisualizationType.TabStops &&
            this.visualizationConfigurationFactory.getConfiguration(visualizationType).testMode !==
                TestMode.Assessments
        );
    }

    private enableToggleOnDetailsViewOpen(test: VisualizationType, tabId: number): void {
        const payload: VisualizationTogglePayload =
            this.createVisualizationTogglePayloadWithNullTelemetry(test);
        this.onVisualizationToggle(payload);
    }

    private createVisualizationTogglePayloadWithNullTelemetry(
        test: VisualizationType,
    ): VisualizationTogglePayload {
        return {
            test,
            enabled: true,
            telemetry: null,
        };
    }

    private onPivotChildSelected = async (
        payload: OnDetailsViewOpenPayload,
        tabId: number,
    ): Promise<void> => {
        this.sidePanelActions.closeSidePanel.invoke('PreviewFeatures');
        this.visualizationActions.updateSelectedPivotChild.invoke(payload);
        await this.detailsViewController
            .showDetailsView(tabId)
            .catch(e => this.logger.error(e.message));
        this.telemetryEventHandler.publishTelemetry(TelemetryEvents.PIVOT_CHILD_SELECTED, payload);
    };

    private onDetailsViewPivotSelected = (payload: OnDetailsViewPivotSelected): void => {
        this.visualizationActions.updateSelectedPivot.invoke(payload);
        this.telemetryEventHandler.publishTelemetry(
            TelemetryEvents.DETAILS_VIEW_PIVOT_ACTIVATED,
            payload,
        );
    };

    private onVisualizationToggle = (payload: VisualizationTogglePayload): void => {
        const telemetryEvent = this.adHocTestTypeToTelemetryEvent[payload.test];
        this.telemetryEventHandler.publishTelemetry(telemetryEvent, payload);
        this.unifiedScanResultActions.startScan.invoke(null);

        if (payload.enabled) {
            this.visualizationActions.enableVisualization.invoke(payload);
        } else {
            this.visualizationActions.disableVisualization.invoke(payload.test);
        }
    };

    private onRescanVisualization = (payload: RescanVisualizationPayload) => {
        this.visualizationActions.disableVisualization.invoke(payload.test);
        this.visualizationActions.enableVisualization.invoke(payload);
        this.unifiedScanResultActions.startScan.invoke(null);
        this.telemetryEventHandler.publishTelemetry(TelemetryEvents.RESCAN_VISUALIZATION, payload);
    };

    private getVisualizationToggleCurrentState = (): void => {
        this.visualizationActions.getCurrentState.invoke(null);
    };

    private getScanResultsCurrentState = (): void => {
        this.visualizationScanResultActions.getCurrentState.invoke(null);
    };

    private onSetHoveredOverSelector = (payload: string[]): void => {
        this.inspectActions.setHoveredOverSelector.invoke(payload);
    };
}
