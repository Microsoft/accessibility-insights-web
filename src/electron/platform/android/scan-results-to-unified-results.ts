// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { InstancePropertyBag, InstanceResultStatus, UnifiedResult } from 'common/types/store-data/unified-data-interface';
import { UUIDGeneratorType } from 'common/uid-generator';
import { DictionaryStringTo } from '../../../types/common-types';
import { RuleInformation } from './rule-information';
import { RuleInformationProvider } from './rule-information-provider';
import { RuleResultsData, ScanResults, ViewElementData } from './scan-results';

export type ConvertScanResultsToUnifiedResultsDelegate = (scanResults: ScanResults, uuidGenerator: UUIDGeneratorType) => UnifiedResult[];

export function convertScanResultsToUnifiedResults(scanResults: ScanResults, uuidGenerator: UUIDGeneratorType): UnifiedResult[] {
    if (!scanResults || !scanResults.ruleResults) {
        return [];
    }

    return createUnifiedResultsFromScanResults(scanResults, uuidGenerator);
}

function createUnifiedResultsFromScanResults(scanResults: ScanResults, uuidGenerator: UUIDGeneratorType): UnifiedResult[] {
    const ruleInformationProvider: RuleInformationProvider = new RuleInformationProvider();
    const viewElementLookup: DictionaryStringTo<ViewElementData> = createViewElementLookup(scanResults);
    const unifiedResults: UnifiedResult[] = [];

    for (const ruleResult of scanResults.ruleResults) {
        const ruleInformation: RuleInformation = ruleInformationProvider.getRuleInformation(ruleResult.ruleId);

        if (ruleInformation) {
            unifiedResults.push(createUnifiedResult(ruleInformation, ruleResult, viewElementLookup, uuidGenerator));
        }
    }

    return unifiedResults;
}

function createViewElementLookup(scanResults: ScanResults): DictionaryStringTo<ViewElementData> {
    const viewElementLookup = {};

    addViewElementAndChildren(viewElementLookup, scanResults.viewElementTree);

    return viewElementLookup;
}

function addViewElementAndChildren(viewElementLookup: DictionaryStringTo<ViewElementData>, element: ViewElementData): void {
    if (element) {
        viewElementLookup[element.axeViewId] = element;
        if (element.children) {
            for (const child of element.children) {
                addViewElementAndChildren(viewElementLookup, child);
            }
        }
    }
}

function createUnifiedResult(
    ruleInformation: RuleInformation,
    ruleResult: RuleResultsData,
    viewElementLookup: DictionaryStringTo<ViewElementData>,
    uuidGenerator: UUIDGeneratorType,
): UnifiedResult {
    return {
        uid: uuidGenerator(),
        ruleId: ruleInformation.ruleId,
        status: getStatus(ruleResult.status),
        descriptors: getDescriptors(viewElementLookup[ruleResult.axeViewId]),
        identifiers: null,
        resolution: {
            'how-to-fix': ruleInformation.howToFix(ruleResult),
        },
    };
}

function getDescriptors(viewElement: ViewElementData): InstancePropertyBag {
    if (viewElement) {
        return {
            className: viewElement.className,
            boundingRectangle: viewElement.boundsInScreen,
            contentDescription: viewElement.contentDescription,
            text: viewElement.text,
        };
    }
    return null;
}

function getStatus(status: string): InstanceResultStatus {
    switch (status) {
        case 'PASS':
            return 'pass';
        case 'FAIL':
            return 'fail';
        default:
            return 'unknown';
    }
}
