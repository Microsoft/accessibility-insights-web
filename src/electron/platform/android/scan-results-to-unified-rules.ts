// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { UnifiedRule } from 'common/types/store-data/unified-data-interface';
import { UUIDGenerator } from 'common/uid-generator';
import { RuleInformation } from './rule-information';
import { RuleInformationProviderType } from './rule-information-provider-type';
import { ScanResults } from './scan-results';

export type ConvertScanResultsToUnifiedRulesDelegate = (
    scanResults: ScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
) => UnifiedRule[];

export function convertScanResultsToUnifiedRules(
    scanResults: ScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
): UnifiedRule[] {
    if (!scanResults) {
        return [];
    }

    const unifiedRules: UnifiedRule[] = [];
    const ruleIds: Set<string> = new Set();

    for (const result of scanResults.ruleResults) {
        const ruleId = result.ruleId;
        if (!ruleIds.has(ruleId)) {
            const ruleInformation = ruleInformationProvider.getRuleInformation(ruleId);

            if (ruleInformation && ruleInformation.includeThisResult(result)) {
                unifiedRules.push(createUnifiedRuleFromRuleResult(ruleInformation, uuidGenerator));
                ruleIds.add(ruleId);
            }
        }
    }

    return unifiedRules;
}

function createUnifiedRuleFromRuleResult(
    ruleInformation: RuleInformation,
    uuidGenerator: UUIDGenerator,
): UnifiedRule {
    return {
        uid: uuidGenerator(),
        id: ruleInformation.ruleId,
        description: ruleInformation.ruleDescription,
        url: null,
        guidance: [],
    };
}
