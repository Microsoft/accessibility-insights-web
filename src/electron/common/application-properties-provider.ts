// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { AppDataAdapter } from 'common/browser-adapters/app-data-adapter';
import { ToolData } from 'common/types/store-data/unified-data-interface';
import { androidAppTitle } from 'content/strings/application';
import { AndroidScanResults } from 'electron/platform/android/android-scan-results';

export type ToolDataDelegate = (scanResults: AndroidScanResults) => ToolData;

export const createGetToolDataDelegate = (appDataAdapter: AppDataAdapter): ToolDataDelegate => {
    return scanResults => {
        return {
            scanEngineProperties: {
                name: 'axe-android',
                version: scanResults.axeVersion,
            },
            applicationProperties: {
                name: androidAppTitle,
                version: appDataAdapter.getVersion(),
            },
        };
    };
};
