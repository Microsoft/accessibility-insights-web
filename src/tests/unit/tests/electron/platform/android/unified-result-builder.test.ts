// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { UUIDGeneratorType } from 'common/uid-generator';
import { ToolDataDelegate } from 'electron/common/application-properties-provider';
import { ScanResults } from 'electron/platform/android/scan-results';
import { ConvertScanResultsToUnifiedResultsDelegate } from 'electron/platform/android/scan-results-to-unified-results';
import { ConvertScanResultsToUnifiedRulesDelegate } from 'electron/platform/android/scan-results-to-unified-rules';
import { createBuilder } from 'electron/platform/android/unified-result-builder';
import { axeRuleResultExample } from 'tests/unit/tests/electron/flux/action-creator/scan-result-example';
import { Mock, MockBehavior } from 'typemoq';

describe('buildUnifiedScanCompletedPayload', () => {
    const exampleScanResults = new ScanResults(axeRuleResultExample);

    it('builds the payload', () => {
        const generateUIDMock = Mock.ofType<UUIDGeneratorType>();

        const getUnifiedResultsMock = Mock.ofType<ConvertScanResultsToUnifiedResultsDelegate>();
        getUnifiedResultsMock
            .setup(converter => converter(exampleScanResults, generateUIDMock.object))
            .returns(() => {
                return [
                    {
                        uid: 'test-result-uid',
                        status: 'pass',
                        ruleId: 'test-rule-id',
                        descriptors: null,
                        identifiers: null,
                        resolution: null,
                    },
                ];
            });

        const getUnifiedRulesMock = Mock.ofType<ConvertScanResultsToUnifiedRulesDelegate>(undefined, MockBehavior.Strict);
        getUnifiedRulesMock
            .setup(converter => converter(exampleScanResults, generateUIDMock.object))
            .returns(() => {
                return [
                    {
                        id: 'test-rule-id',
                        description: 'test-rule-description',
                        url: 'test-url',
                        guidance: [],
                    },
                ];
            });

        const getToolDataMock = Mock.ofType<ToolDataDelegate>();
        getToolDataMock
            .setup(getter => getter(exampleScanResults))
            .returns(() => {
                return {
                    applicationProperties: {
                        name: 'test-app-name',
                        version: 'test-version',
                    },
                    scanEngineProperties: {
                        name: 'test-scan-engine-name',
                        version: 'test-scan-engine-version',
                    },
                };
            });

        const testSubject = createBuilder(
            getUnifiedResultsMock.object,
            getUnifiedRulesMock.object,
            generateUIDMock.object,
            getToolDataMock.object,
        );
        const result = testSubject(exampleScanResults);

        expect(result).toMatchSnapshot();
    });
});
