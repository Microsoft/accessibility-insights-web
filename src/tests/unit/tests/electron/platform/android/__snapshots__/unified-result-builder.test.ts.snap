// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildUnifiedScanCompletedPayload builds the payload 1`] = `
Object {
  "rules": Array [
    Object {
      "description": "test-rule-description",
      "guidance": Array [],
      "id": "test-rule-id",
      "url": "test-url",
    },
  ],
  "scanResult": Array [
    Object {
      "descriptors": null,
      "identifiers": null,
      "resolution": null,
      "ruleId": "test-rule-id",
      "status": "pass",
      "uid": "test-result-uid",
    },
  ],
  "targetAppInfo": Object {
    "name": "com.test.scan-result",
  },
  "toolInfo": Object {
    "applicationProperties": Object {
      "name": "AI-Android",
      "version": "0.0",
    },
    "scanEngineProperties": Object {
      "name": "axe-android",
      "version": "axe-android-test-version",
    },
  },
}
`;
